sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2373,14 +2373,18 @@ class Dataset[T] private[sql](
   @DeveloperApi
   def collectAsArrow(converter: Option[ArrowConverters] = None): ArrowPayload = {
     val cnvtr = converter.getOrElse(new ArrowConverters)
-    withNewExecutionId {
-      try {
-        val collectedRows = queryExecution.executedPlan.executeCollect()
-        cnvtr.internalRowsToPayload(collectedRows, this.schema)
-      } catch {
-        case e: Exception =>
-          throw e
-      }
+    val batchRdd = toArrowBatchBytes()
+    val batches = batchRdd.collect().flatMap(cnvtr.readBatchBytes)
+    new ArrowStaticPayload(batches: _*)
+  }
+
+  def toArrowBatchBytes(): RDD[Array[Byte]] = {
+    val schema_captured = this.schema
+    queryExecution.toRdd.mapPartitions { iter =>
```

Comment:
Thanks, I'll look into that!  We can probably use mapPartitionsInternal because I don't think the closure needs cleaning..

Not sure if it should be moved to SparkPlan, I'll let others weigh in on that.. but it probably shouldn't be public.